cumulus/parachains/runtimes/assets/common/src/lib.rs

```diff
@@ -132,6 +138,185 @@ pub type PoolAssetsConvertedConcreteId<PoolAssetsPalletLocation, Balance> =
+	Matcher: MatchesFungibles<
+		<Runtime as pallet_assets::Config<AssetsInstance>>::AssetId,
+		<Runtime as pallet_assets::Config<AssetsInstance>>::Balance,
+	>,
+	BalanceConverter: ConversionToAssetBalance<
+		u128,
+		<Runtime as pallet_assets::Config<AssetsInstance>>::AssetId,
+		<Runtime as pallet_assets::Config<AssetsInstance>>::Balance,
+	>,
+	AssetsInstance: 'static,
+	<Runtime as pallet_assets::Config<AssetsInstance>>::Balance: Into<u128>,
+{
+	fn convert_asset(asset: &Asset, asset_id: &AssetId) -> Result<Asset, XcmError> {
+		// TODO: Not the best still.
+		let desired_asset: Asset = (asset_id.clone(), 1u128).into(); // To comply with the interface.
+		let (local_asset_id, _) = Matcher::matches_fungibles(&desired_asset).map_err(|error| {
+			log::trace!(
+				target: "xcm::SufficientAssetConverter::convert_asset",
+				"Could not map XCM asset {:?} to FRAME asset. Error: {:?}",
+				asset_id,
+				error,
+			);
+			XcmError::AssetNotFound
+		})?;
+		log::trace!(target: "xcm::SufficientAssetConverter::convert_asset", "local asset id: {:?}", local_asset_id);
+		let Fungibility::Fungible(old_asset_amount) = asset.fun else {
+			log::trace!(
+				target: "xcm::SufficientAssetConverter::convert_asset",
+				"Fee asset is not fungible",
+			);
+			return Err(XcmError::AssetNotFound);
+		};
+		let new_asset_amount = BalanceConverter::to_asset_balance(old_asset_amount, local_asset_id)
+			.map_err(|_| {
+				log::trace!(
+					target: "xcm::SufficientAssetConverter::convert_asset",
+					"Couldn't convert balance of {:?} to balance of {:?}",
+					asset.id,
+					desired_asset.id,
+				);
+				XcmError::TooExpensive
+			})?;
+		let new_asset_amount: u128 = new_asset_amount.try_into().map_err(|_| {
+			log::error!(
+				target: "xcm::SufficientAssetConverter::convert_asset",
+				"Converting balance of {:?} to u128 would overflow",
+				desired_asset.id,
+			);
+			XcmError::Overflow
+		})?;
+		Ok((desired_asset.id.clone(), new_asset_amount).into())
+	}
+
+	fn swap(give: &Asset, want: &Asset) -> Result<Asset, XcmError> {
+		// We need the matcher to select this implementation from the tuple.
+		let _ = Matcher::matches_fungibles(give).map_err(|_| {
+			log::trace!(
+				target: "xcm::SufficientAssetConverter::swap",
+				"Could not map XCM asset {:?} to FRAME asset.",
+				give,
+			);
+			XcmError::AssetNotFound
+		})?;
+		// We don't do a swap, we just returned the asset we need for fees.
+		Ok(want.clone())
+	}
+}
+
+/// Implementation of `AssetConverter` that pays delivery fees by swapping the given asset to the
+/// `Target` asset accepted for fees.
+pub struct SwapAssetConverter<Target, Runtime, Fungibles, Matcher, SwapCredit, AccountId>(
+	PhantomData<(Target, Runtime, Fungibles, Matcher, SwapCredit, AccountId)>,
+);
+impl<Target, Runtime, Fungibles, Matcher, SwapCredit, AccountId> AssetConversion
+	for SwapAssetConverter<Target, Runtime, Fungibles, Matcher, SwapCredit, AccountId>
+where
+	Target: Get<Fungibles::AssetId>,
```

Comment:
Don't we achieve this by having the pools and swapping to it? I don't see the practical use for this and it would complicate things quite a bit. Which asset should you swap to?

Reply:
actually, we do we even need `Target` here? why limit these swaps to just one (or few) asset types?

It could always try to go from `give` to `want` and either succeed or not based on pool existence/liquidity.

Reply:
We want it to swap for the actual asset used for fees. But I guess that could fall to the executor and not to this struct.

Reply:
Yeah, the executor already interrogates the transport protocol about the fees and gets both required id and amount, so when it calls into this it already expects a specific asset back.

The `Target` only "helps" to hard-code the transport fee asset id at runtime config, but I think we can get rid of it.

Reply:
Removed `Target` in the latest version